include/albatross/src/models/patchwork_gp_details.hpp

```diff
@@ -0,0 +1,270 @@
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_MODELS_PATCHWORK_GP_DETAILS_HPP_
+#define ALBATROSS_MODELS_PATCHWORK_GP_DETAILS_HPP_
+
+namespace albatross {
+
+namespace details {
+
+DEFINE_CLASS_METHOD_TRAITS(boundary);
+
+DEFINE_CLASS_METHOD_TRAITS(nearest_group);
+
+DEFINE_CLASS_METHOD_TRAITS(grouper);
+
+/*
+ * Here we make sure that a given class T contains all the required methods
+ * to be PatchworkFunctions and that the types involved are consistent.
+ */
+template <typename T, typename FeatureType>
+class patchwork_functions_are_valid {
+
+  using GroupKey = typename class_method_grouper_traits<
+      T, typename const_ref<FeatureType>::type>::return_type;
+
+  static constexpr bool has_valid_grouper = group_key_is_valid<GroupKey>::value;
+
+  using ConstRefGroupKey = typename const_ref<GroupKey>::type;
+
+  using BoundaryReturnType =
+      typename class_method_boundary_traits<T, ConstRefGroupKey,
+                                            ConstRefGroupKey>::return_type;
+
+  static constexpr bool has_valid_boundary =
+      is_vector<BoundaryReturnType>::value;
+
+  template <typename Key, std::enable_if_t<!std::is_void<Key>::value, int> = 0,
+            typename NearestGroupReturnType =
+                typename class_method_nearest_group_traits<
+                    T, typename const_ref<std::vector<Key>>::type,
+                    ConstRefGroupKey>::return_type>
+  static NearestGroupReturnType nearest_group_test(int);
+
+  template <typename C> static void nearest_group_test(...);
+
+  static constexpr bool has_valid_nearest_group =
+      std::is_same<decltype(nearest_group_test<GroupKey>(0)), GroupKey>::value;
+
+public:
+  static constexpr bool value =
+      has_valid_grouper && has_valid_boundary && has_valid_nearest_group;
+};
+
+} // namespace details
+
+/*
+ * A BoundaryFeature represents a pseudo observation of the difference
+ * between predictions from two different models.  In other words,
+ *
+ *   BoundaryFeature(key_i, key_j, feature)
+ *
+ * represents the quantity
+ *
+ *   model_i.predict(feature) - model_j.predict(feature)
+ *
+ * Patchwork Krigging uses these to force equivalence between two
+ * otherwise independent models.  These are the \delta_{k,l} variables
+ * in Equation 2 from the paper referenced above.
+ */
+template <typename GroupKey, typename FeatureType> struct BoundaryFeature {
+
+  BoundaryFeature(const GroupKey &left_key_, const GroupKey &right_key_,
+                  const FeatureType &feature_)
+      : left_key(left_key_), right_key(right_key_), feature(feature_){};
+
+  BoundaryFeature(GroupKey &&left_key_, GroupKey &&right_key_,
+                  FeatureType &&feature_)
+      : left_key(std::move(left_key_)), right_key(std::move(right_key_)),
+        feature(std::move(feature_)){};
+
+  GroupKey left_key;
+  GroupKey right_key;
```

Comment:
Likewise, can we make this `left_group` and `right_group` or `first_group` and `second_group`?